lib/ansible/modules/files/ini_file.py

```diff
@@ -160,77 +166,140 @@ def do_ini(module, filename, section=None, option=None, value=None,
+
+        try:
+            # line of the first section...
+            first_section_line = sections_start_line[0]
+            section_found = True
+        except IndexError:
+            # ...or, if no section found, the scope to look for the option is the whole file
+            first_section_line = len(ini_lines)
+
+        # Looking for the option
+        option_found = False
+        if first_section_line > 0:
+            # Looking for every line containing this option before the first section
+            # I know, it should be declared only once...
+            option_in_lines = [index for index, line in enumerate(ini_lines[:first_section_line]) if r_key.match(line)]
+            try:
+                # Option found
+                workon_line = option_in_lines[0]
+                option_found = True
+            except IndexError:
+                # option does not exist
+                workon_line = 0
         else:
-            if within_section and option:
-                if state == 'present':
-                    # change the existing option line
-                    if match_opt(option, line):
-                        newline = assignment_format % (option, value)
-                        option_changed = ini_lines[index] != newline
-                        changed = changed or option_changed
-                        if option_changed:
-                            msg = 'option changed'
-                        ini_lines[index] = newline
-                        if option_changed:
-                            # remove all possible option occurrences from the rest of the section
-                            index = index + 1
-                            while index < len(ini_lines):
-                                line = ini_lines[index]
-                                if line.startswith('['):
-                                    break
-                                if match_active_opt(option, line):
-                                    del ini_lines[index]
-                                else:
-                                    index = index + 1
-                        break
-                elif state == 'absent':
-                    # delete the existing line
-                    if match_active_opt(option, line):
-                        del ini_lines[index]
+            # No option before first section or EOF
+            workon_line = 0
+
+        if state == "absent" and option_found > 0:
+            # Delete option only if it has been found
+            ini_lines.pop(workon_line)
+            changed = True
+            msg = 'option removed'
+
+        if state == "present":
+            if option_found:
+                # option found, so it could need, or not, be modified
+
+                # get and prepare existing value to be compared to the target value
+                existing_option = [x.strip() for x in ini_lines[workon_line].split("=")]
```

Comment:
It could be `existing_value = ini_lines[workon_line].split("=")[1]` and `desired_value = value`

Reply:
`existing_option` and `desired_option` are both lists. In each one, it's stored the existing and desired key-value pair, (two items lists), so it's not needed to create more vars to store the value. Maybe names are not the best. I'm leaving the code as it is, but changing the lists names to `existing_option_value` and `desired_option_value`.

Reply:
Ok, but I'm doing nothing with the option as it's already processed, so taking into account option names again it's not needed. I'm changing this as suggested.